### PR TITLE
boards: waveshare: rp2040_zero: Initial support

### PR DESCRIPTION
**Pull request by icpmoles**
Adds initial support for Waveshare RP2040-Zero

### CODE REVIEW

**Review by nordicjm — CHANGES_REQUESTED** [review]
Please read https://docs.zephyrproject.org/latest/contribute/index.html

File: boards/waveshare/rp2040_zero/Kconfig.defconfig [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2024 Iacopo Moles
+# SPDX-License-Identifier: Apache-2.0
+
+
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/Kconfig.defconfig:4]:
```suggestion
```

File: boards/waveshare/rp2040_zero/Kconfig.defconfig [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+if BOARD_RP2040_ZERO
+
+config RP2_FLASH_W25Q080
+	default y
+
+
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/Kconfig.defconfig:10]:
```suggestion
```

File: boards/waveshare/rp2040_zero/Kconfig.defconfig [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+if I2C_DW
+
+config I2C_DW_CLOCK_SPEED
+	default 125
+
+endif #I2C_DW
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/Kconfig.defconfig:16]:
```suggestion
endif # I2C_DW
```

File: boards/waveshare/rp2040_zero/Kconfig.defconfig [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/Kconfig.defconfig:17]:
```suggestion
```

File: boards/waveshare/rp2040_zero/doc/img/RP2040-Zero-1.jpg
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/doc/img/RP2040-Zero-1.jpg:1]: remove whitespace from image, ideally make background transparent, save as webp, then put through https://tinypng.com/

File: boards/waveshare/rp2040_zero/doc/index.rst [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+- Low-power sleep and dormant modes.
+- Drag-and-drop programming using mass storage over USB.
+- 29 × multi-function GPIO pins (20× via edge pinout, others via solder points).
+- 2 × SPI, 2 × I2C, 2 × UART, 4 × 12-bit ADC, 16 × controllable PWM channels.
+- Accurate clock and timer on-chip.
+- Temperature sensor.
+- Accelerated floating-point libraries on-chip.
+- 8 × Programmable I/O (PIO) state machines for custom peripheral support.
+
+
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/doc/index.rst:28]:
```suggestion
```

File: boards/waveshare/rp2040_zero/doc/index.rst [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+     :align: center
+     :alt: RP2040-Zero with pre-soldered pin header
+
+     RP2040-Zero (above) and RP2040-Zero with pre-soldered pin header (below)
+     (Images courtesy of Waveshare)
+
+Supported Features
+==================
+
+The RP2040-Zero board configuration supports the following hardware features:
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/doc/index.rst:44]: Put board name as used with west with \`\` around it and call it board target, not configuration

File: boards/waveshare/rp2040_zero/doc/index.rst [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+- I2C1_SCL : P7
+- SPI0_RX : P16
+- SPI0_CSN : P17
+- SPI0_SCK : P18
+- SPI0_TX : P19
+- ADC_CH0 : P26
+- ADC_CH1 : P27
+- ADC_CH2 : P28
+- ADC_CH3 : P29
+
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/doc/index.rst:119]:
```suggestion
```

File: boards/waveshare/rp2040_zero/doc/index.rst [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+Programming and Debugging
+*************************
+
+Flashing
+========
+
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/doc/index.rst:126]:
```suggestion
```

File: boards/waveshare/rp2040_zero/doc/index.rst [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+Using UF2
+---------
+
+Since it doesn’t expose the SWD pins, you must flash the RP2040-Zero with an UF2 file. By default, building an app for this board will generate a build/zephyr/zephyr.uf2 file. If the RP2040-Zero is powered on with the BOOTSEL button pressed, it will appear on the host as a mass storage device. The UF2 file should be drag-and-dropped to the device, which will flash the RP2040-Zero.
+
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/doc/index.rst:132]:
```suggestion
```

File: boards/waveshare/rp2040_zero/doc/index.rst [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+*************************
+
+Flashing
+========
+
+
+Using UF2
+---------
+
+Since it doesn’t expose the SWD pins, you must flash the RP2040-Zero with an UF2 file. By default, building an app for this board will generate a build/zephyr/zephyr.uf2 file. If the RP2040-Zero is powered on with the BOOTSEL button pressed, it will appear on the host as a mass storage device. The UF2 file should be drag-and-dropped to the device, which will flash the RP2040-Zero.
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/doc/index.rst:131]: `west flash` should also work

File: boards/waveshare/rp2040_zero/doc/index.rst [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+
+Here is an example of building and flashing the sample for driving the built-in RGB led.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/drivers/led/led_strip
+   :board: rp2040_zero
+   :goals: build
+   :compact:
+
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/doc/index.rst:141]:
```suggestion
```

File: boards/waveshare/rp2040_zero/rp2040_zero-pinctrl.dtsi [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+	adc_default: adc_default {
+		group1 {
+			pinmux = <ADC_CH0_P26>, <ADC_CH1_P27>, <ADC_CH2_P28>, <ADC_CH3_P29>;
+			input-enable;
+		};
+	};
+
+	clocks_default: clocks_default {
+	};
+
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/rp2040_zero-pinctrl.dtsi:66]:
```suggestion
```

File: boards/waveshare/rp2040_zero/Kconfig.defconfig [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2024 Iacopo Moles
+# SPDX-License-Identifier: Apache-2.0
+
+
+if BOARD_RP2040_ZERO
+
+config RP2_FLASH_W25Q080
+	default y
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/Kconfig.defconfig:8]: move to `Kconfig` as:
```
config BOARD_RP2040_ZERO
        select RP2_FLASH_W25Q080
```

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: boards/waveshare/rp2040_zero/doc/index.rst (hunk 10 above)
Comment (icpmoles) [reply boards/waveshare/rp2040_zero/doc/index.rst:131]: Can it work without the SWD pins exposed? For [similarly shaped](https://docs.zephyrproject.org/latest/boards/seeed/xiao_rp2040/doc/index.html) rp2040 boards that don't expose the debugger pins the Zephyr documentation only mention UF2 flashing.

Comment (nordicjm) [reply boards/waveshare/rp2040_zero/doc/index.rst:128]: When you have a uf2 runner as the default, west flash will copy the uf2 file over, it has no need for SWD pins, they would only be used with runners that flash using that.

Comment (icpmoles) [reply boards/waveshare/rp2040_zero/doc/index.rst]: Thanks for the explanation. I rewrote the section and now it mentions `west flash` and the manual drag-and-drop.

File: boards/waveshare/rp2040_zero/Kconfig.rp2040_zero [hunk 14]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2024 Iacopo Moles
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_RP2040_ZERO
+	select SOC_RP2040
+
+config BOARD_RP2040_ZERO
```
Comment (nordicjm) [inline boards/waveshare/rp2040_zero/Kconfig.rp2040_zero:7]:
```suggestion
```

**Review by nordicjm — COMMENTED** [review]
Minor nit. and please add a message to the commit to fix compliance issue

**Review by nordicjm — DISMISSED** [review]

File: boards/waveshare/rp2040_zero/rp2040_zero.dts [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+			read-only;
+		};
+
+		/*
+		 * Usable flash. Starts at 0x100, after the bootloader. The partition
+		 * size is 16MB minus the 0x100 bytes taken by the bootloader.
+		 */
+		code_partition: partition@100 {
+			label = "code-partition";
+			reg = <0x100(DT_SIZE_M(16) - 0x100)>;
```
Comment (soburi) [inline boards/waveshare/rp2040_zero/rp2040_zero.dts:51]: missing space?

**Review by soburi — CHANGES_REQUESTED** [review]
code_partition has weird notation.

**Review by soburi — DISMISSED** [review]
LGTM

**Comment by soburi** [comment]
@nordicjm 
Could you revisit this PR?

**Review by nordicjm — DISMISSED** [review]

File: boards/waveshare/rp2040_zero/doc/index.rst [hunk 16]
```diff
@@ -0,0 +1,151 @@
+.. _rp2040_zero:
+
+Waveshare RP2040-Zero
+#####################
```
Comment (kartben) [inline boards/waveshare/rp2040_zero/doc/index.rst:4]:
```suggestion
.. zephyr:board:: rp2040_zero
```

File: boards/waveshare/rp2040_zero/doc/img/RP2040-Zero-M-1.png
Comment (kartben) [inline boards/waveshare/rp2040_zero/doc/img/RP2040-Zero-M-1.png:1]: please drop this image as it is not really adding much information in the context of Zephyr documentation

File: boards/waveshare/rp2040_zero/doc/img/RP2040-Zero-1.png
Comment (kartben) [inline boards/waveshare/rp2040_zero/doc/img/RP2040-Zero-1.png:1]: name the same as board name, i.e. `rp2040_zero.png`

File: boards/waveshare/rp2040_zero/doc/index.rst [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+     :alt: RP2040-Zero
+
+
+.. figure:: img/RP2040-Zero-M-1.png
+     :align: center
+     :alt: RP2040-Zero with pre-soldered pin header
+
+     RP2040-Zero (above) and RP2040-Zero with pre-soldered pin header (below)
+     (Images courtesy of Waveshare)
+
```
Comment (kartben) [inline boards/waveshare/rp2040_zero/doc/index.rst:39]:
```suggestion
```

image will be auto-added by the board directive

**Review by kartben — CHANGES_REQUESTED** [review]

**Review by soburi — DISMISSED** [review]

File: boards/waveshare/rp2040_zero/doc/index.rst (hunk 16 above)
Comment (icpmoles) [reply boards/waveshare/rp2040_zero/doc/index.rst]: I added the `full_name` field in the board.yaml.

Now it works correctly.

**Review by soburi — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Pull request merged**